File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/persist/PeekPersistedExample.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 
Imports System
Imports System.IO
Imports System.Threading
Imports com.db4o

Namespace com.db4odoc.f1.persist
    Public Class PeekPersistedExample
        Inherits Util
        Public Shared Sub main(ByVal args() As String)
            MeasureCarTemperature()
        End Sub

        Public Shared Sub SetObjects()
            File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim car As Car = New Car("BMW")
                db.Set(car)
            Finally
                db.Close()
            End Try
        End Sub

        Public Shared Sub MeasureCarTemperature()
            SetObjects()
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim result As ObjectSet = db.Get(GetType(Car))
                If result.Size() > 0 Then
                    Dim car As Car = CType(result(0), Car)
                    Dim car1 As Car = CType(db.Ext().PeekPersisted(car, 5, True), Car)
                    Dim ch1 As Change1 = New Change1()
                    ch1.Init(car1)
                    Dim car2 As Car = CType(db.Ext().PeekPersisted(car, 5, True), Car)
                    Dim ch2 As Change2 = New Change2()
                    ch2.Init(car2)
                    Thread.Sleep(300)
                    ' We can work on the database object at the same time
                    car.Model = "BMW M3Coupe"
                    db.Set(car)
                    ch1.Kill()
                    ch2.Kill()
                    System.Console.WriteLine("car1 saved to the database: " + db.Ext().IsStored(car1).ToString())
                    System.Console.WriteLine("car2 saved to the database: " + db.Ext().IsStored(car1).ToString())
                    Dim temperature As Integer = CType(((car1.Temperature + car2.Temperature) / 2), Integer)
                    car.Temperature = temperature
                    db.Set(car)
                End If
            Finally
                db.Close()
            End Try
            ÑheckCar()
        End Sub

        Public Shared Sub ÑheckCar()
            Dim db As ObjectContainer = Db4o.OpenFile(Util.YapFileName)
            Try
                Dim result As ObjectSet = db.Get(GetType(Car))
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub
    End Class
End Namespace

File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/persist/PeekPersistedExample.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */
using System;
using System.IO;
using System.Threading;
using com.db4o;

namespace com.db4odoc.f1.persist
{
	public class PeekPersistedExample
	{
		public readonly static string YapFileName = "formula1.yap";

		public static void Main(String[] args) 
		{
			MeasureCarTemperature();
		}
		// end Main
		
		public static void SetObjects()
		{
			File.Delete(YapFileName);
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				Car car = new Car("BMW");
				db.Set(car);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end SetObjects

		public static void MeasureCarTemperature()
		{
			SetObjects();
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet result = db.Get(typeof(Car));
				if (result.Size() > 0)
				{
					Car car = (Car)result[0];
					Car car1  = (Car)db.Ext().PeekPersisted(car, 5, true);
					Change1 ch1 = new Change1();
					ch1.Init(car1);
					Car car2  = (Car)db.Ext().PeekPersisted(car, 5, true);
					Change2 ch2 = new Change2();
					ch2.Init(car2);
					Thread.Sleep(300);
					// We can work on the database object at the same time
					car.Model = "BMW M3Coupe";
					db.Set(car);
					ch1.Stop();
					ch2.Stop();
					System.Console.WriteLine("car1 saved to the database: " + db.Ext().IsStored(car1));
					System.Console.WriteLine("car2 saved to the database: " + db.Ext().IsStored(car1));
					int temperature = (int)((car1.Temperature + car2.Temperature)/2);
					car.Temperature = temperature;
					db.Set(car);
				}
			} 
			finally 
			{
				db.Close();
			}
			ÑheckCar();
		}
		// end MeasureCarTemperature
	
		public static void ÑheckCar()
		{
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet result = db.Get(typeof(Car));
				ListResult(result);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end ÑheckCar

		public static void ListResult(ObjectSet result)
		{
			Console.WriteLine(result.Count);
			foreach (object item in result)
			{
				Console.WriteLine(item);
			}
		}
		// end ListResult
	}
}
